extension/entitystore/ec2Info.go

```diff
@@ -0,0 +1,144 @@
+
+	// region is used while making call to describeTags Ec2 API for AutoScalingGroup
+	Region string
+
+	metadataProvider ec2metadataprovider.MetadataProvider
+	logger           *zap.Logger
+	done             chan struct{}
+}
+
+func (ei *EC2Info) initEc2Info() {
+	ei.logger.Debug("Initializing EC2Info")
+	if err := ei.setInstanceIDAccountID(); err != nil {
+		return
+	}
+	if err := ei.setAutoScalingGroup(); err != nil {
+		return
+	}
+	ei.logger.Debug("Finished initializing EC2Info")
+	ei.ignoreInvalidFields()
+}
+
+func (ei *EC2Info) setInstanceIDAccountID() error {
+	for {
+		metadataDoc, err := ei.metadataProvider.Get(context.Background())
+		if err != nil {
+			ei.logger.Warn("Failed to get Instance ID / Account ID through metadata provider", zap.Error(err))
+			wait := time.NewTimer(1 * time.Minute)
+			select {
+			case <-ei.done:
+				wait.Stop()
+				return errors.New("shutdown signal received")
+			case <-wait.C:
+				continue
+			}
+		}
+		ei.logger.Debug("Successfully retrieved Instance ID and Account ID")
+		ei.InstanceID = metadataDoc.InstanceID
+		ei.AccountID = metadataDoc.AccountID
+		return nil
+	}
+}
+
+func (ei *EC2Info) setAutoScalingGroup() error {
+	retry := 0
+	for {
+		var waitDuration time.Duration
+		if retry < len(ec2tagger.BackoffSleepArray) {
+			waitDuration = ec2tagger.BackoffSleepArray[retry]
+		} else {
+			waitDuration = ec2tagger.BackoffSleepArray[len(ec2tagger.BackoffSleepArray)-1]
+		}
+
+		wait := time.NewTimer(waitDuration)
+		select {
+		case <-ei.done:
+			wait.Stop()
+			return errors.New("shutdown signal received")
+		case <-wait.C:
+		}
+
+		if retry > 0 {
+			ei.logger.Debug("Initial retrieval of tags and volumes", zap.Int("retry", retry))
+		}
+
+		if err := ei.retrieveAsgName(); err != nil {
+			ei.logger.Warn("Unable to fetch instance tags with imds", zap.Int("retry", retry), zap.Error(err))
+		} else {
+			ei.logger.Debug("Retrieval of auto-scaling group tags succeeded")
+			return nil
+		}
+
+		retry++
+	}
+
+}
+
+func (ei *EC2Info) retrieveAsgName() error {
+	tags, err := ei.metadataProvider.InstanceTags(context.Background())
+	if err != nil {
+		ei.logger.Debug("Failed to get tags through metadata provider", zap.Error(err))
+		return err
+	} else if strings.Contains(tags, ec2tagger.Ec2InstanceTagKeyASG) {
+		asg, err := ei.metadataProvider.InstanceTagValue(context.Background(), ec2tagger.Ec2InstanceTagKeyASG)
+		if err != nil {
+			ei.logger.Error("Failed to get AutoScalingGroup through metadata provider", zap.Error(err))
+		} else {
+			ei.logger.Debug("AutoScalingGroup retrieved through IMDS")
+			ei.AutoScalingGroup = asg
+		}
+	}
+	return nil
+}
+
+func newEC2Info(metadataProvider ec2metadataprovider.MetadataProvider, done chan struct{}, region string, logger *zap.Logger) *EC2Info {
+	return &EC2Info{
+		metadataProvider: metadataProvider,
+		done:             done,
+		Region:           region,
+		logger:           logger,
+	}
+}
+
+func (ei *EC2Info) ignoreInvalidFields() {
```

Comment:
Do we have this check in the `ec2tagger`? Is there a reason we have this as a separate function and not as part of the `set` functions?

Reply:
This check is not in `ec2tagger` but was mentioned as a security risk for the component. No specific reason for separate function but either way should be fine. Let me know if any concerns about this.

Reply:
Add a commit to change it in the set function

Reply:
Non-blocking, but already implemented so will add in.